src/plugins/vis_type_xy/public/vis_component.tsx

```diff
@@ -0,0 +1,299 @@
+    return Boolean(filters.length);
+  };
+
+  const toggleLegend = useCallback(() => {
+    setShowLegend((value) => {
+      const newValue = !value;
+      if (props.uiState?.set) {
+        props.uiState.set('vis.legendOpen', newValue);
+      }
+      return newValue;
+    });
+  }, [props.uiState]);
+
+  const setColor = useCallback(
+    (newColor: string | null, seriesLabel: string | number, event: BaseSyntheticEvent) => {
+      if ((event as KeyboardEvent).key && (event as KeyboardEvent).key !== keys.ENTER) {
+        return;
+      }
+
+      const colors = props.uiState?.get('vis.colors') || {};
+      if (colors[seriesLabel] === newColor || !newColor) {
+        delete colors[seriesLabel];
+      } else {
+        colors[seriesLabel] = newColor;
+      }
+
+      if (props.uiState?.set) {
+        props.uiState.setSilent('vis.colors', null);
+        props.uiState.set('vis.colors', colors);
+        props.uiState.emit('colorChanged');
+      }
+    },
+    [props.uiState]
+  );
+
+  const { visData, visParams } = props;
+  const config = getConfig(visData, visParams);
+  const timeZone = getTimeZone();
+  const xDomain =
+    config.xAxis.scale.type === ScaleType.Ordinal ? undefined : getXDomain(config.aspects.x.params);
+  const hasBars = visParams.seriesParams.some(
+    ({ type, data: { id: paramId } }) =>
+      type === ChartType.Histogram &&
+      config.aspects.y.find(({ aggId }) => aggId === paramId) !== undefined
+  );
+  const adjustedXDomain =
+    config.xAxis.scale.type === ScaleType.Ordinal
+      ? undefined
+      : getAdjustedDomain(visData.rows, config.aspects.x, timeZone, xDomain, hasBars);
+  const legendPosition = useMemo(() => config.legend.position ?? Position.Right, [
+    config.legend.position,
+  ]);
+  const isDarkMode = getThemeService().useDarkMode();
+  const getSeriesName = getSeriesNameFn(config.aspects, config.aspects.y.length > 1);
+  const nonStackedBars = visParams.seriesParams.filter(
+    ({ type, data: { id: paramId }, valueAxis: groupId, mode }) => {
+      const barAspect =
+        type === ChartType.Histogram && config.aspects.y.find(({ aggId }) => aggId === paramId);
+
+      if (!barAspect) {
+        return false;
+      }
+
+      const yAxisScale = config.yAxes.find(({ groupId: axisGroupId }) => axisGroupId === groupId)
+        ?.scale;
+      return !(mode === 'stacked' || yAxisScale?.mode === 'percentage');
+    }
+  );
+  const hasNonStackedBars = nonStackedBars.length > 1;
+
+  const getSeriesColor = useCallback(
+    (series: XYChartSeriesIdentifier) => {
+      const seriesName = getSeriesName(series);
+      if (!seriesName) {
+        return;
+      }
+
+      const overwriteColors: Record<string, string> = props.uiState?.get
+        ? props.uiState.get('vis.colors', {})
+        : {};
+
+      allSeries.push(seriesName);
+      return getColorsService().createColorLookupFunction(allSeries, overwriteColors)(seriesName);
```

Comment:
This logic will look pretty different when switching over to the palette service - we will need the total number of series (including dynamic splits) and the position of the current series along with the name.

I'm aware this will get added later, just to make sure it won't be super complicated to compute this information

Reply:
Yeah the dynamic splits are the big issue here, since elastic charts is now performing the split on the data, we don't know all the series names ahead of time. Unfortunately, I think this is a necessary evil to match the existing vislib color service until we transition to use the new palettes, at which time we could hopefully do away with this code.